source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -461,8 +472,11 @@ Network::FilterStatus Filter::establishUpstreamConnection() {
     // cluster->trafficStats()->upstream_cx_none_healthy in the latter case.
     getStreamInfo().setResponseFlag(StreamInfo::ResponseFlag::NoHealthyUpstream);
     onInitFailure(UpstreamFailureReason::NoHealthyUpstream);
+    return Network::FilterStatus::StopIteration;
   }
-  return Network::FilterStatus::StopIteration;
+  // Allow OnData() to receive data before connect if so configured
+  return receive_before_connect_ ? Network::FilterStatus::Continue
```

Comment:
In the case of `receive_before_connect_`, now that earlier filters have seen whatever data they need and allowed the filter chain to continue, should tcp_proxy `readDisable(true)` to prevent additional data from getting buffered while the upstream connection is established?

If not, then I think there needs to be some limit to the data enforced. One option is to leave it in the Connection buffer so that flow control can be triggered.